features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,242 @@
+# Objective
+
+- Provide cached result to Chat Container API calls
+- Compatible to old API interface
+
+# Overview
+
+Container contains a local store for caching resources. The local store caches resources by resource id.
+
+### Store Update
+
+When SDK recieves result from server, for each record, the newly fetched record would always replace the old one with the same record id if there is one.
+
+### API
+
+For fetch API of each resource, there would be a new callback that should return result from cache immediately.
+
+The interface for each language may be different.
+
+```
+getResource(params, callback, cachedCallback?)
+```
+
+Resource includes conversation and message.
+
+The cachedCallback should be optional, thus compatible to the old API interace.
+
+### Expected UI Display / Common Use Case
+
+Client App is expected to keep a single list of messages for display in table. And they are required to keep track to the newest and oldest fetched messages to determine when to fetch result from server.
+
+Client App should replace old message with new message if the same message id is found.
+
+Deleted messages are also returned in fetch API, with empty content and marked as deleted. This is to keep the whole message list complete. Client App is expected to skip these messages in UI display.
+
+The conversation view controller in UI Kit would use the cached version of chat API. Thus, developer should be able to use this new feature easily.
+
+The controller should have the following behaviour:
+
+- keep a cursor pointing to a message or the latest message, set a window size of how many messages should be kept in memory
+- when moving the cursor around (scrolling around or jumping around), it should erase messages outside the window and fetch messages newly entered the window
+- fetch the messages to fill the window when enter view, if the cursor is not set, it would stick to the latest message
+- subscribe to new messages, but only append to bottom if the window is on the latest message side
+- display, allow resend, cancel pending sent message, i.e. send by user but not responded by server yet
+- sync latest messages on reconnect if the window is on the latest message side
+
+# Sample Code for Conversation List
```

Comment:
Any sample code for message list?

Reply:
And also sample code for createMessage

Reply:
Shall we also include how to handle when pubsub is connected, the users should get fetchMessage_after-time?